Initial Module Buildout

dannyvink opened the pull request:
This exposes the function `dockerComposeToApi(yaml|object)` that returns an object that is properly formatted for triggering `createService`

dannyvink committed:
Initial commit

dannyvink committed:
Add Volume support

dannyvink committed:
Add file comment

dannyvink committed:
Add port validation

dannyvink committed:
Rename full to name

dannyvink committed:
Add test for object

ecwillis commented on lib/index.js:
```diff
@@ -0,0 +1,195 @@
+    spec.TaskTemplate.ContainerSpec['Env'] = [];
+
+    // Convert { test: 1 } to 'test=1' format.
+    for (var key in input[name]['environment']) {
```
I think we would also need support for `environment` being passed like this:
```
environment:
   - 'NODE_ENV=production'
```
Which we do roughly half the time.

jgallen23 replied:
lets support both ways

ecwillis commented on lib/index.js:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  if ('labels' in input[name]) {
+    spec.TaskTemplate.ContainerSpec['Labels'] = {};
```
In our service stacks, these are located at `deploy.labels` like so:
```
deploy:
    labels:
         label: 'value'
```

There are also nodes: `deploy.restart_policy`, `deploy.placement`, `deploy.replicas`, `deploy.update_config`, and `deploy.resources`

ecwillis commented on lib/index.js:
```diff
@@ -0,0 +1,195 @@
+    }
+    spec.TaskTemplate['RestartPolicy'] = policy;
+  }
+
```
Seems to be missing `networks`.

ecwillis reviewed (changes requested):
I think we need to review service stacks.

jgallen23 commented on lib/index.js:
```diff
@@ -0,0 +1,195 @@
+
+  if ('volumes' in input[name] && input[name]['volumes'].length > 0) {
+    spec.TaskTemplate.ContainerSpec['Mounts'] = [];
+    for (var i = 0; i < input[name]['volumes'].length; i++) {
```
one other thing is `.reduce` and/or `.map` might make a lot of this code cleaner

jgallen23 commented on lib/index.js:
```diff
@@ -0,0 +1,195 @@
+    }
+  }
+
+  if ('volumes' in input[name] && input[name]['volumes'].length > 0) {
```
wonder if it would be cleaner to put all of these in a `./lib/volumes.js` file that way the file doesn't just keep growing.  then you could loop over the object and just call require(`./lib/${key}`)(obj)

jgallen23 replied:
and throw an error if something is there that isn't supported

jgallen23 commented:
might have missed, but we need placement constraints and resources too

dannyvink committed:
Break out into separate files

dannyvink committed:
Break out into separate files

dannyvink committed:
Add resources

dannyvink committed:
Add validation and networks

dannyvink committed:
Remove incorrect config/secrets setup, add validation

dannyvink committed:
Update dependencies

dannyvink commented:
Broke everything out into separate files and added the requested missing fields! If I missed anything, let me know!

I have concerns regarding secrets / configs. I originally implemented them incorrectly and the conversion from docker-compose to docker-engine isn't quite 1 to 1 so I removed those fields for now. Any thoughts there would be appreciated @jgallen23 @ecwillis

ecwillis commented on lib/index.js:
```diff
@@ -0,0 +1,110 @@
+
+  let validation = Joi.validate(input, serviceSchema);
+  if (validation.error) {
+    console.log(validation.error);
```
See asana and https://gist.github.com/ecwillis/0a5afaee897a3bbb5c2037144b01aeba for comments and example doc.

dannyvink committed:
Reduce validation restrictions
